spark/src/main/java/org/apache/iceberg/actions/BaseActions.java

```diff
@@ -22,24 +22,16 @@
 import org.apache.iceberg.Table;
 import org.apache.spark.sql.SparkSession;
 
-public class Actions {
+public abstract class BaseActions {
 
   private SparkSession spark;
   private Table table;
 
-  private Actions(SparkSession spark, Table table) {
+  protected BaseActions(SparkSession spark, Table table) {
```

Comment:
Is there a simpler way to split this between Spark 2 and 3?

What about extending `Actions` and then selecting whether `Spark2Actions` or `Spark3Actions` will be used at runtime using reflection? Then these two static methods would have to change, but none of the test could would. Anything that is only implemented in Spark 3 would throw `UnsupportedOperationException` here and override the method in `Spark3Actions`.

I think that might significantly reduce the amount of changes needed for this.

Reply:
That seems like a more complicated solution to me and exposes runtime errors to the user directly. I would much rather library users have compile time safety.

In general I usually try to avoid reflection for things like this, especially if we can't guarantee the phblic api will actually be fulfilled.

If this was scala we could do some implicits here that would give us compile time selectivity but I don't think there is a pure Java way to do the same...

Reply:
Did a quick impl of your suggestion in case you wanted to discuss the differences between the two with concrete examples. https://github.com/apache/iceberg/pull/1616